Metricbeat Kubernetes module and metricsets docs proposal

eedugon opened the pull request:
As agreed with @exekias , I'm adding in this PR some changes in Kubernetes module main document and in the majority of metricsets, including descriptions and configuration details for each one.

eedugon commented:
@exekias , please review this PR. I'm going to need some help with some links, and I'm not totally sure that I have modified the right file.
I'm doubting between the file I have modified (`metricbeat/docs/modules/kubernetes.asciidoc`) and (`metricbeat/module/kubernetes/_meta/docs.asciidoc`)

elasticmachine commented:
Pinging @elastic/integrations-platforms (Team:Platforms)

exekias commented:
@ChrsMark could you have a look to this one and help @eedugon to move it forward :pray: ?

ChrsMark commented:
Hey @eedugon! The change must be applied on `metricbeat/module/kubernetes/_meta/docs.asciidoc` and then run `mage update` under `metricbeat` directory so as to automatically update  `metricbeat/docs/modules/kubernetes.asciidoc`.

eedugon commented:
@ChrsMark , @exekias , I think i have fixed the CLA issue and I have made the changes into the right file.
As you said `mage update` also changed the final file too. This latest commit includes the content after running the update too.
Let me know your thoughts.

This would require also updating 7.x docs, but maybe it's better to review them in master (`docs-preview` doesn't work for me), and as soon as all looks great we can do the backport.

ChrsMark commented on metricbeat/docs/modules/kubernetes.asciidoc:
```diff
@@ -14,11 +14,46 @@ This module fetches metrics from Kubernetes several components:
-All metricsets with the `state_` prefix require `hosts` field pointing to kube-state-metrics
-service within the cluster, while the rest should be pointed to kubelet service. Check the
-example configuration on how to do it.
+Some of the previous components are running in each of the Kubernetes nodes (like `kubelet` or `proxy`) while others provide a single cluster-wide endpoint. This is important to determine the optimal configuration and running strategy for the different metricsets included in the module.
```
```suggestion
Some of the previous components are running on each of the Kubernetes nodes (like `kubelet` or `proxy`) while others provide a single cluster-wide endpoint. This is important to determine the optimal configuration and running strategy for the different metricsets included in the module.
```

ChrsMark commented on metricbeat/docs/modules/kubernetes.asciidoc:
```diff
@@ -14,11 +14,46 @@ This module fetches metrics from Kubernetes several components:
+
+Kubernetes module is a bit complex as its internal metricsets require access to a wide variety of endpoints.
+
+This section highlights and introduces some groups of metricsets with similar endpoint access needs. For more details on the metricsets see the ****configuration example link*** and ***metricsets reference link**** sections below.
```
configuration example link?
metricsets reference link?

Should those be replaced by sth else?

eedugon replied:
@ChrsMark , I'd like to link there to the `Example Configuration` and `Metricsets` sections at the end of the final document. Take a look at the current https://www.elastic.co/guide/en/beats/metricbeat/current/metricbeat-module-kubernetes.html

How should I add those links?

eedugon replied:
Removing links and resolving conversation :)

ChrsMark commented on metricbeat/docs/modules/kubernetes.asciidoc:
```diff
@@ -14,11 +14,46 @@ This module fetches metrics from Kubernetes several components:
+
+==== node / system / pod / container / module
+
+The default metricsets `container`, `node`, `pod`, `system` and `volume` require access to the `kubelet endpoint` in each of the Kubernetes nodes, hence it's recommended to include them as part of a `Metricbeat DaemonSet` or standalone Metricbeats running in the hosts.
```
```suggestion
The default metricsets `container`, `node`, `pod`, `system` and `volume` require access to the `kubelet endpoint` on each of the Kubernetes nodes, hence it's recommended to include them as part of a `Metricbeat DaemonSet` or standalone Metricbeats running in the hosts.
```

ChrsMark commented on metricbeat/docs/modules/kubernetes.asciidoc:
```diff
@@ -14,11 +14,46 @@ This module fetches metrics from Kubernetes several components:
+- Create `Kubernetes Services` to make `kube-controller-manager` and `kube-scheduler` available and configure the metricsets to point to these services as part of a Metricbeat `Deployment`.
+- Use `Autodiscovery` functionality as part of a Metricbeat DaemonSet and include the metricsets in a conditional template applied for the specific pods.
+
+Note: In some "As a Service" Kubernetes implementations, like `GKE`, the master nodes or even the pods running in the masters won't be visible. In these cases it won't be possible to use `scheduler` nor `controllermanager` metricsets.
```
```suggestion
Note: In some "As a Service" Kubernetes implementations, like `GKE`, the master nodes or even the pods running on the masters won't be visible. In these cases it won't be possible to use `scheduler` nor `controllermanager` metricsets.
```

ChrsMark commented on metricbeat/module/kubernetes/_meta/docs.asciidoc:
```diff
@@ -7,11 +7,46 @@ This module fetches metrics from Kubernetes several components:
-All metricsets with the `state_` prefix require `hosts` field pointing to kube-state-metrics
-service within the cluster, while the rest should be pointed to kubelet service. Check the
-example configuration on how to do it.
+Some of the previous components are running in each of the Kubernetes nodes (like `kubelet` or `proxy`) while others provide a single cluster-wide endpoint. This is important to determine the optimal configuration and running strategy for the different metricsets included in the module.
```
```suggestion
Some of the previous components are running on each of the Kubernetes nodes (like `kubelet` or `proxy`) while others provide a single cluster-wide endpoint. This is important to determine the optimal configuration and running strategy for the different metricsets included in the module.
```

ChrsMark commented on metricbeat/module/kubernetes/_meta/docs.asciidoc:
```diff
@@ -7,11 +7,46 @@ This module fetches metrics from Kubernetes several components:
+
+==== node / system / pod / container / module
+
+The default metricsets `container`, `node`, `pod`, `system` and `volume` require access to the `kubelet endpoint` in each of the Kubernetes nodes, hence it's recommended to include them as part of a `Metricbeat DaemonSet` or standalone Metricbeats running in the hosts.
```
```suggestion
The default metricsets `container`, `node`, `pod`, `system` and `volume` require access to the `kubelet endpoint` in each of the Kubernetes nodes, hence it's recommended to include them as part of a `Metricbeat DaemonSet` or standalone Metricbeats running on the hosts.
```

ChrsMark commented on metricbeat/module/kubernetes/_meta/docs.asciidoc:
```diff
@@ -7,11 +7,46 @@ This module fetches metrics from Kubernetes several components:
+- Create `Kubernetes Services` to make `kube-controller-manager` and `kube-scheduler` available and configure the metricsets to point to these services as part of a Metricbeat `Deployment`.
+- Use `Autodiscovery` functionality as part of a Metricbeat DaemonSet and include the metricsets in a conditional template applied for the specific pods.
+
+Note: In some "As a Service" Kubernetes implementations, like `GKE`, the master nodes or even the pods running in the masters won't be visible. In these cases it won't be possible to use `scheduler` nor `controllermanager` metricsets.
```
```suggestion
Note: In some "As a Service" Kubernetes implementations, like `GKE`, the master nodes or even the pods running on the masters won't be visible. In these cases it won't be possible to use `scheduler` nor `controllermanager` metricsets.
```

ChrsMark commented:
Left some minor comments. Content looks good overall! I also restarted the failed docs job.

elasticmachine commented:
## :broken_heart: Build Failed 

<!-- BUILD BADGES-->
[![Pipeline View](https://img.shields.io/badge/pipeline-pipeline%20-green)](https://beats-ci.elastic.co/blue/organizations/jenkins/Beats%2Fbeats-beats-mbp%2FPR-17636/detail/PR-17636/10//pipeline) [![Test View](https://img.shields.io/badge/test-test-green)](https://beats-ci.elastic.co/blue/organizations/jenkins/Beats%2Fbeats-beats-mbp%2FPR-17636/detail/PR-17636/10//tests) [![Changes](https://img.shields.io/badge/changes-changes-green)](https://beats-ci.elastic.co/blue/organizations/jenkins/Beats%2Fbeats-beats-mbp%2FPR-17636/detail/PR-17636/10//changes) [![Artifacts](https://img.shields.io/badge/artifacts-artifacts-yellow)](https://beats-ci.elastic.co/blue/organizations/jenkins/Beats%2Fbeats-beats-mbp%2FPR-17636/detail/PR-17636/10//artifacts) [![preview](https://img.shields.io/badge/docs-preview-yellowgreen)](http://beats_17636.docs-preview.app.elstc.co/diff) [![stats](https://img.shields.io/badge/stats-stats%20-blue)](https://ela.st/observabtl-ci-stats)

<!-- BUILD SUMMARY-->
<details><summary>Expand to view the summary</summary>
<p>

#### Build stats

* Build Cause: [Pull request #17636 updated]
* Start Time: 2020-04-29T12:49:40.384+0000

* Duration: 28 min 15 sec (1694997)


* Commit: b810a61296b1eba381c8d4350b8c13851789e2ee




</p>
</details>

<!-- TEST RESULTS IF ANY-->


<!-- STEPS ERRORS IF ANY-->


### Steps errors

  <details><summary>Expand to view the steps failures</summary>
  <p>
  
  * **Name**: Make check
    * Description: make check
    * Result: FAILURE
    
    * Duration: 7 min 52 sec<
    
    * Start Time: 2020-04-29T13:10:54.619+0000
    
  
  </p>
  </details>



<!-- LOG OUTPUT-->
### Log output

<details><summary>Expand to view the last 100 lines of log output</summary>
<p>

```
[2020-04-29T13:17:48.905Z] + FILE=heartbeat/build/coverage/full.cov
[2020-04-29T13:17:48.905Z] + [ -f heartbeat/build/coverage/full.cov ]
[2020-04-29T13:17:48.905Z] + FILE=libbeat/build/coverage/full.cov
[2020-04-29T13:17:48.905Z] + [ -f libbeat/build/coverage/full.cov ]
[2020-04-29T13:17:48.905Z] + FILE=metricbeat/build/coverage/full.cov
[2020-04-29T13:17:48.905Z] + [ -f metricbeat/build/coverage/full.cov ]
[2020-04-29T13:17:48.905Z] + FILE=packetbeat/build/coverage/full.cov
[2020-04-29T13:17:48.905Z] + [ -f packetbeat/build/coverage/full.cov ]
[2020-04-29T13:17:48.905Z] + FILE=winlogbeat/build/coverage/full.cov
[2020-04-29T13:17:48.905Z] + [ -f winlogbeat/build/coverage/full.cov ]
[2020-04-29T13:17:48.905Z] + FILE=journalbeat/build/coverage/full.cov
[2020-04-29T13:17:48.905Z] + [ -f journalbeat/build/coverage/full.cov ]
[2020-04-29T13:17:50.586Z] Stage "Build and Test" skipped due to earlier failure(s)
[2020-04-29T13:17:50.861Z] Stage "Elastic Agent x-pack" skipped due to earlier failure(s)
[2020-04-29T13:17:50.863Z] Stage "Elastic Agent x-pack Windows" skipped due to earlier failure(s)
[2020-04-29T13:17:50.864Z] Stage "Elastic Agent Mac OS X" skipped due to earlier failure(s)
[2020-04-29T13:17:50.865Z] Stage "Filebeat oss" skipped due to earlier failure(s)
[2020-04-29T13:17:50.867Z] Stage "Filebeat x-pack" skipped due to earlier failure(s)
[2020-04-29T13:17:50.867Z] Stage "Filebeat Mac OS X" skipped due to earlier failure(s)
[2020-04-29T13:17:50.868Z] Stage "Filebeat Windows" skipped due to earlier failure(s)
[2020-04-29T13:17:50.869Z] Stage "Heartbeat" skipped due to earlier failure(s)
[2020-04-29T13:17:50.870Z] Stage "Auditbeat oss" skipped due to earlier failure(s)
[2020-04-29T13:17:50.871Z] Stage "Auditbeat x-pack" skipped due to earlier failure(s)
[2020-04-29T13:17:50.872Z] Stage "Libbeat" skipped due to earlier failure(s)
[2020-04-29T13:17:50.872Z] Stage "Libbeat x-pack" skipped due to earlier failure(s)
[2020-04-29T13:17:50.873Z] Stage "Metricbeat OSS Unit tests" skipped due to earlier failure(s)
[2020-04-29T13:17:50.874Z] Stage "Metricbeat OSS Integration tests" skipped due to earlier failure(s)
[2020-04-29T13:17:50.875Z] Stage "Metricbeat Python integration tests" skipped due to earlier failure(s)
[2020-04-29T13:17:50.876Z] Stage "Metricbeat x-pack" skipped due to earlier failure(s)
[2020-04-29T13:17:50.877Z] Stage "Metricbeat crosscompile" skipped due to earlier failure(s)
[2020-04-29T13:17:50.878Z] Stage "Metricbeat Mac OS X" skipped due to earlier failure(s)
[2020-04-29T13:17:50.879Z] Stage "Metricbeat Windows" skipped due to earlier failure(s)
[2020-04-29T13:17:50.881Z] Stage "Packetbeat" skipped due to earlier failure(s)
[2020-04-29T13:17:50.882Z] Stage "dockerlogbeat" skipped due to earlier failure(s)
[2020-04-29T13:17:50.883Z] Stage "Winlogbeat" skipped due to earlier failure(s)
[2020-04-29T13:17:50.883Z] Stage "Winlogbeat Windows x-pack" skipped due to earlier failure(s)
[2020-04-29T13:17:50.885Z] Stage "Functionbeat" skipped due to earlier failure(s)
[2020-04-29T13:17:50.886Z] Stage "Journalbeat" skipped due to earlier failure(s)
[2020-04-29T13:17:50.887Z] Stage "Generators" skipped due to earlier failure(s)
[2020-04-29T13:17:50.888Z] Stage "Kubernetes" skipped due to earlier failure(s)
[2020-04-29T13:17:51.070Z] Stage "Heartbeat" skipped due to earlier failure(s)
[2020-04-29T13:17:51.072Z] Stage "Auditbeat oss" skipped due to earlier failure(s)
[2020-04-29T13:17:51.073Z] Stage "Libbeat" skipped due to earlier failure(s)
[2020-04-29T13:17:51.074Z] Stage "Packetbeat" skipped due to earlier failure(s)
[2020-04-29T13:17:51.075Z] Stage "dockerlogbeat" skipped due to earlier failure(s)
[2020-04-29T13:17:51.076Z] Stage "Winlogbeat" skipped due to earlier failure(s)
[2020-04-29T13:17:51.077Z] Stage "Functionbeat" skipped due to earlier failure(s)
[2020-04-29T13:17:51.078Z] Stage "Journalbeat" skipped due to earlier failure(s)
[2020-04-29T13:17:51.079Z] Stage "Generators" skipped due to earlier failure(s)
[2020-04-29T13:17:51.686Z] Failed in branch Elastic Agent x-pack
[2020-04-29T13:17:51.687Z] Failed in branch Elastic Agent x-pack Windows
[2020-04-29T13:17:51.691Z] Failed in branch Elastic Agent Mac OS X
[2020-04-29T13:17:51.696Z] Failed in branch Filebeat oss
[2020-04-29T13:17:51.697Z] Failed in branch Filebeat x-pack
[2020-04-29T13:17:51.698Z] Failed in branch Filebeat Mac OS X
[2020-04-29T13:17:51.699Z] Failed in branch Filebeat Windows
[2020-04-29T13:17:51.700Z] Failed in branch Auditbeat x-pack
[2020-04-29T13:17:51.702Z] Failed in branch Libbeat x-pack
[2020-04-29T13:17:51.703Z] Failed in branch Metricbeat OSS Unit tests
[2020-04-29T13:17:51.704Z] Failed in branch Metricbeat OSS Integration tests
[2020-04-29T13:17:51.705Z] Failed in branch Metricbeat Python integration tests
[2020-04-29T13:17:51.706Z] Failed in branch Metricbeat x-pack
[2020-04-29T13:17:51.708Z] Failed in branch Metricbeat crosscompile
[2020-04-29T13:17:51.709Z] Failed in branch Metricbeat Mac OS X
[2020-04-29T13:17:51.709Z] Failed in branch Metricbeat Windows
[2020-04-29T13:17:51.710Z] Failed in branch Winlogbeat Windows x-pack
[2020-04-29T13:17:51.711Z] Failed in branch Kubernetes
[2020-04-29T13:17:52.014Z] Stage "Heartbeat" skipped due to earlier failure(s)
[2020-04-29T13:17:52.015Z] Stage "Auditbeat oss" skipped due to earlier failure(s)
[2020-04-29T13:17:52.017Z] Stage "Libbeat" skipped due to earlier failure(s)
[2020-04-29T13:17:52.018Z] Stage "Winlogbeat" skipped due to earlier failure(s)
[2020-04-29T13:17:52.020Z] Stage "Functionbeat" skipped due to earlier failure(s)
[2020-04-29T13:17:52.021Z] Stage "Generators" skipped due to earlier failure(s)
[2020-04-29T13:17:52.154Z] Failed in branch Packetbeat
[2020-04-29T13:17:52.155Z] Failed in branch dockerlogbeat
[2020-04-29T13:17:52.156Z] Failed in branch Journalbeat
[2020-04-29T13:17:52.416Z] Stage "Heartbeat" skipped due to earlier failure(s)
[2020-04-29T13:17:52.418Z] Stage "Auditbeat oss" skipped due to earlier failure(s)
[2020-04-29T13:17:52.420Z] Stage "Libbeat" skipped due to earlier failure(s)
[2020-04-29T13:17:52.421Z] Stage "Functionbeat" skipped due to earlier failure(s)
[2020-04-29T13:17:52.422Z] Stage "Generators" skipped due to earlier failure(s)
[2020-04-29T13:17:52.455Z] Failed in branch Winlogbeat
[2020-04-29T13:17:52.789Z] Failed in branch Heartbeat
[2020-04-29T13:17:52.790Z] Failed in branch Libbeat
[2020-04-29T13:17:52.790Z] Failed in branch Functionbeat
[2020-04-29T13:17:52.791Z] Stage "Auditbeat oss" skipped due to earlier failure(s)
[2020-04-29T13:17:52.793Z] Stage "Generators" skipped due to earlier failure(s)
[2020-04-29T13:17:52.932Z] Failed in branch Auditbeat oss
[2020-04-29T13:17:52.933Z] Failed in branch Generators
[2020-04-29T13:17:54.279Z] Running on Jenkins in /var/lib/jenkins/workspace/Beats_beats-beats-mbp_PR-17636
[2020-04-29T13:17:54.571Z] [INFO] getVaultSecret: Getting secrets
[2020-04-29T13:17:54.713Z] Masking supported pattern matches of $VAULT_ADDR or $VAULT_ROLE_ID or $VAULT_SECRET_ID
[2020-04-29T13:17:55.657Z] + chmod 755 generate-build-data.sh
[2020-04-29T13:17:55.657Z] + ./generate-build-data.sh https://beats-ci.elastic.co/blue/rest/organizations/jenkins/pipelines/Beats/beats-beats-mbp/PR-17636/ https://beats-ci.elastic.co/blue/rest/organizations/jenkins/pipelines/Beats/beats-beats-mbp/PR-17636/runs/10 FAILURE 1694997
[2020-04-29T13:17:56.208Z] INFO: curl https://beats-ci.elastic.co/blue/rest/organizations/jenkins/pipelines/Beats/beats-beats-mbp/PR-17636/runs/10/steps/?limit=10000 -o steps-info.json
[2020-04-29T13:17:57.119Z] INFO: curl https://beats-ci.elastic.co/blue/rest/organizations/jenkins/pipelines/Beats/beats-beats-mbp/PR-17636/runs/10/tests/?status=FAILED -o tests-errors.json
[2020-04-29T13:17:57.119Z] jq: error (at tests-errors.json:4): Cannot index string with string "_links"
[2020-04-29T13:17:57.119Z] jq: error (at tests-errors.json:4): Cannot index string with string "_class"
[2020-04-29T13:17:57.119Z] jq: error (at tests-errors.json:4): Cannot index string with string "state"
[2020-04-29T13:17:57.119Z] jq: error (at tests-errors.json:4): Cannot index string with string "hasStdLog"

```

</p>
</details>



<!--PIPELINE
{"commit":"98ebd5738594ef83f0c659b1137152bf785972f6","number":"10","status":"FAILURE","url":"https://beats-ci.elastic.co/job/Beats/job/beats-beats-mbp/job/PR-17636/10/"}
PIPELINE-->

ChrsMark commented on metricbeat/docs/modules/kubernetes.asciidoc:
```diff
@@ -14,11 +14,46 @@ This module fetches metrics from Kubernetes several components:
+Some of the previous components are running on each of the Kubernetes nodes (like `kubelet` or `proxy`) while others provide a single cluster-wide endpoint. This is important to determine the optimal configuration and running strategy for the different metricsets included in the module.
 
-The default metricsets are `container`, `node`, `pod`, `system` and `volume`.
+For a complete reference on how to configure and run this module on Kubernetes as part of a `DaemonSet` and a `Deployment`, there's a complete example manifest available in <<running-on-kubernetes,Running Metricbeat on Kubernetes>> document.
```
nit 
```suggestion
For a complete reference on how to configure and run this module on Kubernetes as part of a `DaemonSet` and a `Deployment`, there's a complete example manifest available in <<running-on-kubernetes, Running Metricbeat on Kubernetes>> document.
```

ChrsMark commented on metricbeat/docs/modules/kubernetes.asciidoc:
```diff
@@ -14,11 +14,46 @@ This module fetches metrics from Kubernetes several components:
+All metricsets with the `state_` prefix require `hosts` field pointing to `kube-state-metrics
+service` within the cluster. As the service provides cluster-wide metrics, there's no need to fetch them per node, hence the recommendation is to run these metricsets as part of a `Metricbeat Deployment` with one only replica.
+
+Note: Kube-state-metrics is not deployed by default in Kubernetes. For these cases the instructions for its deployment are available [here](https://github.com/kubernetes/kube-state-metrics#kubernetes-deployment). Generally `kube-state-metrics` runs a `Deployment` and is accessible via a service called `kube-state-metrics` on `kube-system` namespace, which will be the service to use in our configu ...
```
Urls are provided like this:
```suggestion
Note: Kube-state-metrics is not deployed by default in Kubernetes. For these cases the instructions for its deployment are available https://github.com/kubernetes/kube-state-metrics#kubernetes-deployment[here]. Generally `kube-state-metrics` runs a `Deployment` and is accessible via a service called `kube-state-metrics` on `kube-system` namespace, which will be the service to use in our configuration.
```

eedugon replied:
Changed in my local branch and pushed

ChrsMark commented on metricbeat/docs/modules/kubernetes.asciidoc:
```diff
@@ -14,11 +14,46 @@ This module fetches metrics from Kubernetes several components:
+
+==== scheduler and controllermanager
+
+These metricsets require access to the Kubernetes `controller-manager` and `scheduler` endpoints. By default, these pods run only in the master node, and they are not exposed via a Service, but there are different strategies available for its configuration:
```
```suggestion
These metricsets require access to the Kubernetes `controller-manager` and `scheduler` endpoints. By default, these pods run only on master nodes, and they are not exposed via a Service, but there are different strategies available for its configuration:
```

eedugon replied:
Changed in my local branch and pushed

ChrsMark commented on metricbeat/module/kubernetes/apiserver/_meta/docs.asciidoc:
````diff
@@ -1 +1,32 @@
+  - get
+```
+
+The previous configuration and RBAC requirement is available in the complete example manifest proposed in <<running-on-kubernetes,Running Metricbeat on Kubernetes>> document.
````
```suggestion
The previous configuration and RBAC requirement is available in the complete example manifest proposed in <<running-on-kubernetes, Running Metricbeat on Kubernetes>> document.
```

ChrsMark commented on metricbeat/module/kubernetes/apiserver/_meta/docs.asciidoc:
````diff
@@ -1 +1,32 @@
+Configuration example using https and token based authentication:
+
+```yaml
+    - module: kubernetes
````
Can you we fix the indentation here? It should be one shift left.

eedugon replied:
pushed and solved

eedugon replied:
ops... I forgot the docs update.... wait for it :)
